### DISCUSSION TITLE
mutate vs trigger

### DISCUSSION THREAD

**Discussion by mdrach** [vercel/swr#2050] [Q&A] [answered]
[dumb question alert] I'm trying to understand the difference between `trigger` in the [new `useSWRMutation` hook](https://github.com/vercel/swr/pull/1450/files#diff-b18b8a22a555ef7d7f2f4fb45b01f28ec28a074127469247a3daae77e25628fb) which replaces `mutate` from `useSWR`.

When updating `useSWR` -> `useSWRMutation` in the [error handling example](https://swr.vercel.app/examples/error-handling), `trigger` appears to just be [a name change](https://codesandbox.io/s/exciting-ives-rvxuh2?file=/src/App.tsx) to the uninformed eye.  Can someone explain the improvement that `trigger` provides over `mutate` here?

**Comment by swand0g** [comment, marked as answer]
I think the big thing is that `useSWRMutation` does not initialize the request to the resource automatically, and there is no `revalidateOnFocus` and the like. If you used the new hook, you would have to manually trigger it before it ever fetched any data. So I think it might not be right to think of it as an improvement in that scenario, since it's actually there to trigger the first fetch of data, as opposed to `mutate` and `useSWR` for which data is always automatically fetched.

It's also tied in to `useSWRMutation`'s concept of being able to update remote data (eg POST and PUT), whereas `mutate` is mainly meant to refetch a resource (eg `/api/user` and GET) or update it's local value, not actually trigger an update to its value on a database or server or something

**Reply by artemis-prime** [reply]
Please clarify what trigger() does.  Your wording assumes a lot of existing familiarity.

**Comment by vaynevayne** [comment]
I have the same doubt

i find useSwrMutation is better,   useSWRMutation 's key is only url,
but, useSwr's need all params series as key, it same as react-query, It's hard to pre-declare all request parameters, it's too rigid, I've been using react-query, and it's killing me

https://swr.vercel.app/zh-CN/docs/arguments

```jsx
useSWR('/api/user', url => fetchWithToken(url, token)) //  ❌
const { data: user } = useSWR(['/api/user', token], ([url, token]) => fetchWithToken(url, token)) //  ✅
```

```jsx
import useSWRMutation from 'swr/mutation'
 
// 实现 fetcher
// 额外的参数可以通过第二个参数 `arg` 传入
// 在下例中，`arg` 为 `'my_token'`
async function updateUser(url, { arg }: { arg: string }) {
  await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${arg}`
    }
  })
}
 
function Profile() {
  // 一个类似 useSWR + mutate 的 API，但是它不会自动发送请求
  const { trigger } = useSWRMutation('/api/user', updateUser, options?)
 
  return <button onClick={() => {
    // 以特定参数触发 `updateUser` 函数
    trigger('my_token')
  }}>Update User</button>
}

```
I did not directly ask and answer your question about mutate vs trigger, because I am not familiar with it either, but I can definitely suggest that you should not continue to use useSwr, otherwise you will experience the same pain as I do, and the conclusion is to choose trigger.
